### PR TITLE
[GPU] Fix Select kernel compilation failure on broadcast inputs with different ranks

### PR DESCRIPTION
**Pull request by t8**
### Details

The Select GPU kernel (`select_gpu_ref.cl`) fails to compile when input tensors have different dimensionalities — specifically when a lower-rank condition tensor (e.g., 4D mask) is broadcast to a higher-rank output (e.g., 5D). This commonly occurs with PyTorch-traced `.masked_fill()` or `.tril()` operations where a 2D boolean mask is broadcast across batch/head/spatial dimensions.

**Root cause:** `select_gpu_ref.cl` defines `INPUT_0`, `INPUT_1`, `INPUT_2` macros using `OUTPUT_DIMS` to determine how many index arguments to pass to each input's `GET_INDEX_SAFE` macro. But `GET_INDEX_SAFE` is generated per-tensor based on that tensor's actual layout (in `jitter.cpp`), so a 4D input's `GET_INDEX_SAFE` accepts 4 arguments while the 5D output's `#if OUTPUT_DIMS == 5` block passes 5. This causes `"too many arguments provided to function-like macro invocation"`.

**Fix:** Move `INPUT_N` macro generation from the `.cl` template into `select_kernel_base.cpp`, where each input's actual dimensionality (`Dimentions()`) is used to select the correct index arguments.

**Changes:**
- `select_kernel_base.cpp`: Generate `INPUT_0`, `INPUT_1`, `INPUT_2` JIT constants with per-input dimensionality (4D → `b,f,y,x`, 5D → `b,f,z,y,x`, 6D → `b,f,w,z,y,x`)
- `select_gpu_ref.cl`: Remove hardcoded `#if OUTPUT_DIMS` blocks for `INPUT_N` definitions (now JIT-defined)
- `select_gpu_test.cpp`: Add regression test for 5D output with 4D broadcast mask

**Reproducer:**

```python
import torch, torch.nn as nn, openvino as ov

class M(nn.Module):
    def __init__(self):
        super().__init__()
        self.register_buffer('mask', torch.triu(torch.ones(16, 16, dtype=torch.bool), diagonal=1))
    def forward(self, x):
        return x.masked_fill(self.mask, 0.0)

model = M().eval()
x = torch.randn(1, 4, 2, 16, 16)  # 5D — FAILS before fix
traced = torch.jit.trace(model, (x,))
ov_model = ov.convert_model(traced, example_input=(x,))
compiled = ov.Core().compile_model(ov_model, "GPU")  # ProgramBuilder crash
```

Ranks 2D–4D work; 5D fails with MASK macro expansion error. After this fix, all ranks compile correctly.

### Tickets

- #35271

### AI Assistance
- AI assistance used: yes
- AI was used to draft the fix and reproducer script. Human validation: build verified on Linux (48-core server), all 37 existing non-f16 select GPU tests pass with zero regressions, new regression test passes, fix verified end-to-end on Intel Core Ultra X7 358H iGPU (Panther Lake) via WSL2 — 5D Select ops that previously crashed now compile and execute correctly.

### CODE REVIEW

File: src/plugins/intel_gpu/src/kernel_selector/kernels/select/select_kernel_base.cpp [hunk 1]
```diff
@@ -41,6 +41,22 @@ JitConstants SelectKernelBase::GetJitConstantsCommon(const select_params& params
 
     jit.AddConstant(MakeJitConstant("INPUTS_DECLS", inputs_decls));
 
+    // Generate INPUT_N macros with correct dimensionality for each input tensor.
```
Comment (Lyamin-Roman) [inline src/plugins/intel_gpu/src/kernel_selector/kernels/select/select_kernel_base.cpp:44]: [random spot] Shouldn't all inputs be broadcasted to the output rank in `canonicalize_shapes(..)`?
src/plugins/intel_gpu/src/graph/impls/ocl/select.cpp:46

Comment (t8) [reply src/plugins/intel_gpu/src/kernel_selector/kernels/select/select_kernel_base.cpp:44]: (Looked into this with AI):

Good question. `static_canonicalize_shapes()` does call `extend_shape_to_rank_from_begin()` on all inputs, which should pad them to `max(4, output_rank)`. And it works -- for static shapes:

| Condition shape | Value shape | Static/Dynamic | GPU compile |
|---|---|---|---|
| [16, 16] (2D) | [1, 4, 2, 16, 16] (static) | static | OK |
| [1, 1, 16, 16] (4D) | [1, 4, 2, 16, 16] (static) | static | OK |
| [16, 16] (2D) | [1, ?, ?, ?, ?] (dynamic) | dynamic | **FAILS** |
| [1, 1, 16, 16] (4D) | [1, ?, ?, ?, ?] (dynamic) | dynamic | **FAILS** |

The bug is specific to the **dynamic shape (shape-agnostic) path**. In `primitive_base.hpp:98-101`:

```cpp
auto kernel_params = ImplType::get_kernel_params(ImplType::static_canonicalize_shapes(impl_param));
kernel_params.is_shape_agnostic = impl_param.is_dynamic();
kernel_params.set_dynamic_shape_offsets();
```

`static_canonicalize_shapes` correctly extends the ranks, but when `is_shape_agnostic = true`, the JIT generation for `GET_INDEX_SAFE` macros appears to use the original input ranks rather than the canonicalized ones. This is why the bug only reproduces with models from `ov.convert_model()` (which produces dynamic shapes) but not with the C++ unit test (which uses static shapes).

Our kernel-level fix (generating `INPUT_N` per input's actual `Dimentions()`) works regardless of which JIT path is taken, since it queries the rank at JIT-generation time rather than relying on the `.cl` template's `OUTPUT_DIMS` assumption. The deeper fix would be to ensure the shape-agnostic JIT path preserves the canonicalized ranks for `GET_INDEX_SAFE` macro generation.

Happy to adjust the approach if you'd prefer the fix in the canonicalization/shape-agnostic path instead.

Reproducer for the static vs dynamic comparison: https://gist.github.com/t8/3b2d826b0aeb0e748b41568f8c15a249

Comment (Lyamin-Roman) [reply src/plugins/intel_gpu/src/kernel_selector/kernels/select/select_kernel_base.cpp:44]: >but when is_shape_agnostic = true, the JIT generation for GET_INDEX_SAFE macros appears to use the original input ranks rather than the canonicalized ones

It is not clear where the difference lies between these behaviors, since the shape-agnostic kernels/dynamic case should not affect the rank of tensors and `GET_INDEX_SAFE` processing (src/plugins/intel_gpu/src/kernel_selector/jitter.cpp `DataTensorJitConstant::GetDefinitions()`)
If it is possible to find and solve the global problem of inconsistency between the update logic of the `GET_DATA_INDEX`  JIT constants, then it is better to fix it globally
As far as I understand, the generation of constants for the kernel begins after its selection src/plugins/intel_gpu/src/graph/impls/ocl/primitive_base.hpp
`create(..)` -> `auto best_kernel = kernel_selector.get_best_kernel(kernel_params);`
That uses `kernel_params` with already canonicalized shapes

Comment (t8) [reply src/plugins/intel_gpu/src/kernel_selector/kernels/select/select_kernel_base.cpp:44]: Apologies, you are right. Spent the morning testing further and found the reproducer was running on prebuilt OV 2026.0.0. This bug was fixed in #34537 (included starting from 2026.1.0). After rebuilding from master I can no longer reproduce this issue. 

Appreciate the patience. Closing this now!
